A51-xds-custom-backend-metrics.md

```diff
@@ -0,0 +1,382 @@
+* Utilization metrics show the server status machine wide, independent of any RPC.
+
+### Wire Format
+
+Backend metric data will be encoded in the ORCA format for transmission on the wire. 
+Specifically, it will be transmitted as a binary-encoded [LoadReport](https://github.com/cncf/udpa/blob/04548b0d99d4e70b29310ebccc8e01f2deeed43a/udpa/data/orca/v1/orca_load_report.proto#L14) protobuf.
+
+### Per-Request Metrics Reporting
+
+#### Wire Format
+For metrics reported on a per-request basis, the ORCA LoadReport protobuf will be encoded in the 
+trailing metadata sent by the backend.  The metadata key will be `endpoint-load-metrics-bin`.
+
+Note that if the backend application adds this data to the trailing metadata of each RPC, 
+it will be sent regardless of whether the client actually needs it.  
+This could conceivably lead to wasted bandwidth, CPU, and memory if the backend is sending data 
+that the client does not need, but this case should be fairly rare; 
+backends will generally not be written to populate data that they do not need.  
+Also, we want to avoid the complexity of introducing some kind of negotiation for the client to 
+tell the backend that it wants this data.
+
+#### The Client API
+LB policies will have the ability to access the load report data 
+as part of intercepting the trailing metadata sent from the server.
+Note that the API must be structured such that if multiple LB policies are interested in the data (e.g., both xDS and WRR),
+the protobuf deserialization needs to happen only once. The exact API for this may differ across languages, 
+but here are a couple of possible approaches:
+* The deserialized data can be attached to the metadata element, so that it can be deserialized once and accessed multiple times.
+* The deserialized data can be stored in the call data and accessed by the LB policy code when it intercepts the trailing metadata.
+
+In addition, we also discussed the possibility of having the LB policy get a separate callback for the load report data, 
+distinct from the callback it gets to intercept trailing metadata.  
+However, we think this is not a good idea, since it is likely to cause ordering problems
+(there is no guarantee of the order in which the callbacks would be seen by the LB policy) 
+and may incur additional synchronization overhead.
+
+#### The Server API 
+Using a server interceptor, it is convenient to attach the metrics as part of the
+trailing metadata of the server call and send back with the RPC response.
+
+Define a metric recorder that is unique per RPC. Server applications should obtain a reference and 
+use it to record metrics specifically for the current RPC call. Implementing this may leverage a 
+[context](https://grpc.github.io/grpc-java/javadoc/io/grpc/Context.html) or similar mechanism, 
+and a language should implement one if not already existed.
+The metric recoder looks like this:
```

Comment:
Instead of having Java-specific API info here, please describe the API in a language-agnostic way here, and move the Java API specification to an appendix (which we should also include for the other languages).